### ISSUE TITLE
Install camlp5 with ocamlfind

### ISSUE THREAD

**Issue by bschommer** [camlp5/camlp5#26]
Since coq requires camlp5 to be installed via ocamlfind it would be nice if this would be done during the installation, in a way that avoids relying on opam magic.

**Comment by roglo** [comment]
I had indeed the problem. Talking with @Zimmi48, we found that the problem is that camlp5 library was installed under the ocaml library instead of its subdirectory site-lib. I changed that a few days ago and now findlib finds camlp5.

**Comment by roglo** [comment]
There is a problem, @ejgallego seems not to agree with this change.

**Comment by ejgallego** [comment]
Most distributions seems to be building `findlib` with `-sitelib` set to OCaml std library directory so indeed I think that 3519ba0ae1238844bafe62c351e276f0e6d1b038 has broken the install of `camlp5` in Debian and OPAM.

Indeed I think the change is pointless in that case, but this is entirely a subjective measure.

@bschommer did not provide any detail at all of their setup, distribution, ocaml version, etc... so it is hard to guess what the problem is.

**Comment by bschommer** [comment]
Sorry for the missing details, in my setup I did not install `findlib` with `-sitelib` set to the OCaml std library but rather let it be the default. I think one way around this all would be to just install the `META` file in the correct folder which is my workaround in this case and which seems to work just fine.

**Comment by bschommer** [comment]
I think the easiest solution would be to replace the line:
```
	cp META "$(DESTDIR)$(LIBDIR)/$(CAMLP5N)/."
```
in [etc/Makefile](https://github.com/camlp5/camlp5/blob/master/etc/Makefile) with
```
        ocamlfind install camlp5 META
```

**Comment by roglo** [comment]
It would require to install ocamlfind before installing Camlp5.

**Comment by bschommer** [comment]
Maybe add a configure switch for this?

**Comment by ejgallego** [comment]
I think that the general consensus among packagers is that the non-standard `site-lib` directory is something unwanted and thus they disable it.

I guess in this case you could just set's camlp5 `libdir` appropiatedly, but configuring findlib in a more standard way seems sensible.

**Comment by roglo** [comment]
Ok, I am going to revert this change. For the moment, I just saw that a new version 4.08.0 of ocaml is in preparation and I am updating Camlp5 to accept it.

**Comment by roglo** [comment]
Problem: I don't know how to configure findlib. I tried
  ./configure -sitelib $(ocamlc -where):$(ocamlc -where)/site-lib
Since I installed Camlp5 back to $(ocamlc -where)
But after compilation and installation
  $ ocamlfind -query camlp5
  ocamlfind: Package `camlp5' not found

**Comment by ejgallego** [comment]
That's a typical configuration:
```
"./configure" "-bindir" bin "-sitelib" lib "-mandir" man "-config" "%{lib}%/findlib.conf" "-no-custom" "-no-topfind" {preinstalled}
```

**Comment by Zimmi48** [comment]
> Most distributions seems to be building findlib with -sitelib set to OCaml std library directory so indeed I think that 3519ba0 has broken the install of camlp5 in Debian and OPAM.

IMHO if packagers decide to change the value of `sitelib` it is their problem to fix the install phase of the packages as well. camlp5 has a `-libdir` option that could be used. It doesn't make sense that the default configuration of camp5 would not work with the default version of findlib.

**Comment by ejgallego** [comment]
I wonder how many OCaml packages do work OOTB with the `findlib` defaults; I'd bet not many as it uses a really odd choice of path. Anyways, @bschommer problem can be fixed without any patch to `camlp5`.

Whether to change camlp5 or not, I dunno; personally I wouldn't do it as it seems that today's findlib choice is considered as a bit of an anomaly; it would also break most camlp5 packages, so there is potential for some backslash. But yeah, if "consistency" is desired then go for it.

**Comment by roglo** [comment]
There is indeed a problem: if camlp5 library was installed under ocaml library, then when we install it under ocaml-library/site-lib/camlp5, the old files of ocaml-library/camlp5 may remain.

To be clean, I could delete that possible directory (ocaml-library/camlp5) at installation. Then, the packagers see an error if they still relied on the fact that camlp5 library is there, so they know that something has to be changed in their camlp5 installations.

**Comment by bschommer** [comment]
I agree with @ejgallego that there is nothing to fix here and I would not change camlp5 at the moment. The only thing I would propose is to just document somewhere that if camlp5 is used in combination with ocamlfind one needs to take care that it `libdir` is installed in `sitelib`.

**Comment by roglo** [comment]
No: if camlp5 is installed in site-lib, there is nothing to change in findlib.

**Comment by bschommer** [comment]
I meant that if one wants to use camlp5 in combination with ocamlfind then it should be installed in the `sitelib` directory of ocamlfind, be it the default one `lib/ocaml/site-lib` or as it is the case for example for opam `lib/ocaml`.

**Comment by roglo** [comment]
I don't know what to do... The problem comes from the change in Coq: now, it searches camlp5 through ocamlfind. It works if camlp5 have been installed with opam, but not if you don't use opam. I think the solution is indeed to install camlp5 now in site-lib like I did. Am I wrong?

**Comment by ejgallego** [comment]
> The problem comes from the change in Coq: now, it searches camlp5 through ocamlfind.

Sorry, the problem is orthogonal to what Coq does.

> I think the solution is indeed to install camlp5 now in site-lib like I did. Am I wrong?

I think the previous default was more sensible given that this is what 99% of ocaml packages do.

> It works if camlp5 have been installed with opam, but not if you don't use opam.

It works if you properly set your findlib / camlp5.

I think the most sensible thing to do is to keep the old default, as @bschommer has mentioned this change would break everything, and indeed most distributions are correcting the dubious choice of default path in findlib.

**Comment by roglo** [comment]
Ouais, j'ai fait ça finalement. Je me configure camlp5 avec son -libdir $(ocamlc -where)/site-lib et j'arrive à compiler Coq. Par défaut c'est comme avant.

**Comment by ejgallego** [comment]
Ça va, mais vraiment le plus practique est compiler `findlib` avec `site-lib` sans suffix.

**Comment by ejgallego** [comment]
De ça façon, ton instalation serai compatible avec toutes les versions de Coq.

**Comment by roglo** [comment]
L'option site-lib de findlib ne marche pas, elle ne permet pas de spécifier plusieurs chemins, car ça bugue dans la configuration si on fait ça. Et si on précise uniquement le chemin de camlp5 alors il ne trouve plus les autres !

**Comment by ejgallego** [comment]
Il faut spécifier le chemin du base, normalement /lib/ocaml. Regarde ce que OPAM fais.

**Comment by ejgallego** [comment]
Concrètement https://github.com/ocaml/opam-repository/blob/master/packages/ocamlfind/ocamlfind.1.8.0/opam#L8
